docs/development/design/build-images.rst

```diff
@@ -0,0 +1,216 @@
+
+* ``ubuntu20-nopdf`` (from ``ubuntu20-base``)
+  * same as ``ubuntu20`` but based on ``ubuntu20-base`` instead
+
+.. note::
+
+   I don't think it's useful to have ``ubuntu20-py37`` exposed to users,
+   since the Python version is selected by using the config file's ``python.version`` keyword,
+   we only update patch versions and we don't remove them (unless together with OS changes).
+
+.. Build all these images with Docker
+   docker build -t readthedocs/build:ubuntu20-base -f Dockerfile.base .
+   docker build -t readthedocs/build:ubuntu20-nopdf -f Dockerfile.nopdf .
+   docker build -t readthedocs/build:ubuntu20-pdf -f Dockerfile.pdf .
+   docker build -t readthedocs/build:ubuntu20 -f Dockerfile .
+
+   Check the shared space between images
+   docker system df --verbose | grep -E 'SHARED SIZE|readthedocs'
+
+
+Custom images
+-------------
+
+There are some dependencies that are not easy to update and keep compatibility with all the users at the same time.
+Upgrading ``nodejs`` may make lot of old projects expecting the older version to start failing all their builds.
+On the other hand, sticking with an old version avoid users requiring a newer version to build their documentation.
+To handle this case and others, we have been thinking on supporting custom Docker images.
+
+It's not clear to me how it would be the implementation of this, but I see different paths to discuss and explore:
+
+#. Allow a ``build.dockerfile`` config pointing to a ``Dockerfile``
+  * ``FROM readthedocs/build:ubuntu20`` is required to be a valid image (to share layers)
+  * the image is build each time a build is triggered consuming build time
+#. Create a branch per custom image in ``readthedocs-docker-images`` repository
+  * use ``ubuntu20`` as base image and add the custom extra requirements
+  * build the image using our current process (Docker Hub)
+  * add the custom image to our ``-ops`` repository
+  * re-build builders to pull down the new custom image
+  * set the project to use this custom image, eg. ``readthedocs/build:<project-slug>``
+
+
+Updating versions over time
+---------------------------
+
+How do we add/upgrade a Python version?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Python patch versions can be upgraded and backported to all the images without problems.
+There is only needed to rebuild ``ubuntu20`` and most of the layers will remain shared with ``-base`` and ``-pdf``.
+
+In case we need to *add* a new Python version, the situation is similar.
+We can add the new version by using ``pyenv`` and rebuilding the ``ubuntu20`` image.
+
+
+How do we upgrade system versions?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We usually don't upgrade these dependencies unless we upgrade the Ubuntu version.
+So, they will be only upgraded when we go from Ubuntu 18.04 LTS to Ubuntu 20.04 LTS for example.
+
+Examples of these versions are:
+
+* doxygen
+* git
+* subversion
+* pandoc
+* nodejs / npm
```

Comment:
Should we use `nodenv` (similar to `pyenv`) and install a specific (multiple?) version/s of node/npm? We were already requested a way to install a different version of node some time ago and I created this isseu https://github.com/readthedocs/readthedocs-docker-images/issues/107.

cc @ericholscher

Reply:
I think we should probably use nodeenv to at least install a specific version. I don't know if we need multiple right now, but using nodeenv will make it easier if we need to.